src/odemis/acq/align/autofocus.py

```diff
@@ -587,10 +591,258 @@ def estimateAutoFocusTime(detector, scanner=None, steps=MAX_STEPS_NUMBER):
+
+        logging.debug("Turn on the light")
+        # Turn on the light and check it is on
+        bl = model.getComponent(role="brightlight")
+        f = turnOnLightAndCheck(bl, dets[0])
+        try:
+            f.result(timeout=60)
+        except TimeoutError:
+            f.cancel()
+            logging.warning("Light doesn't appear to turn on after 60s")
+        if future._autofocus_state == CANCELLED:
+            logging.warning("Autofocus procedure is cancelled after turning on the light")
+            raise CancelledError()
+        future._actions_time.pop(0)
+        future.set_progress(end=time.time() + sum(future._actions_time))
+
+        logging.debug("Adjust the optical path")
+        # Go to the special focus mode
+        if align_mode == "lens-align":
+            opath = "spec-focus"
+        elif align_mode == "fiber-align":
+            opath = "spec-fiber-focus"
+        # change the optical path
+        fopm = opm.setPath(opath)
+        fopm.result()
+        if future._autofocus_state == CANCELLED:
+            logging.warning("Autofocus procedure is cancelled after closing the slit")
+            raise CancelledError()
+        future._actions_time.pop(0)
+        future.set_progress(end=time.time() + sum(future._actions_time))
+
+        # Autofocus is not manual
+        if start_autofocus == True:
+            # Configure each detector with good settings
+            for d in dets:
+                # The stream takes care of configuring its detector, so no need
+                if model.hasVA(d, "binning"):
+                    d.binning.value = d.binning.clip((2, 2))
+                if model.hasVA(d, "exposureTime"):
+                    d.exposureTime.value = d.exposureTime.clip(0.2)
+            ret = {}
+            logging.debug("Run AutoFocusSpectrometer")
+            try:
+                # run AutoFocusSpectrometer
+                logging.debug("AutoFocusSpectrometer starts")
+                try:
+                    future._running_subf = AutoFocusSpectrometer(spgr, focuser, dets, selector, streams)
+                    ret = future._running_subf.result(timeout=100)
+                except TimeoutError:
+                    f.cancel()
+                    logging.warning("Timeout error for autofocus spectrometer")
+            except IOError:
+                if future._autofocus_state == CANCELLED:
+                    raise CancelledError()
+                raise
+            if future._autofocus_state == CANCELLED:
+                logging.warning("Autofocus procedure is cancelled after the completion of the autofocus of the spectrometer")
+                raise CancelledError()
+            future._actions_time.pop(0)
+            future.set_progress(end=time.time() + sum(future._actions_time))
+
+            logging.debug("Revert the optical path")
+            # revert the path
+            if align_mode == "lens-align":
+                # # Ensure the latest image shows the slit focused
```

Comment:
What you need to do here is:
1.  Take the first stream (of `streams`). If there is no such stream => you're done!
2.  Ensure it's started (with `_playStream(s.detector, s)`)
3.  Wait one image (with `s.detector.data.get(asap=False)`)
4.  Stop the stream.